tests/unit_tests/dft/include/compute_tester.hpp

```diff
@@ -0,0 +1,120 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef ONEMKL_COMPUTE_TESTER_HPP
+#define ONEMKL_COMPUTE_TESTER_HPP
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+#include "oneapi/mkl.hpp"
+#include "test_helper.hpp"
+#include "test_common.hpp"
+#include "reference_dft.hpp"
+
+template <oneapi::mkl::dft::precision precision, oneapi::mkl::dft::domain domain>
+struct DFT_Test {
+    using descriptor_t = oneapi::mkl::dft::descriptor<precision, domain>;
+
+    template <typename ElemT>
+    using usm_allocator_t = sycl::usm_allocator<ElemT, sycl::usm::alloc::shared, 64>;
+
+    using PrecisionType =
+        typename std::conditional_t<precision == oneapi::mkl::dft::precision::SINGLE, float,
+                                    double>;
+
+    using InputType = typename std::conditional_t<domain == oneapi::mkl::dft::domain::REAL,
+                                                  PrecisionType, std::complex<PrecisionType>>;
+    using OutputType = std::complex<PrecisionType>;
+
+    enum class TestType { buffer, usm };
+
+    const std::int64_t size;
+    const std::int64_t conjugate_even_size;
+    static constexpr int error_margin = 10;
+
+    sycl::device *dev;
+    sycl::queue sycl_queue;
+    sycl::context cxt;
+
+    std::vector<InputType> input;
+    std::vector<PrecisionType> input_re;
+    std::vector<PrecisionType> input_im;
+    std::vector<OutputType> out_host_ref;
+
+    DFT_Test(sycl::device *dev, std::int64_t size)
+            : dev{ dev },
+              size{ static_cast<std::int64_t>(size) },
+              conjugate_even_size{ 2 * (size / 2 + 1) },
+              sycl_queue{ *dev, exception_handler },
+              cxt{ sycl_queue.get_context() } {
+        input = std::vector<InputType>(size);
+        input_re = std::vector<PrecisionType>(size);
+        input_im = std::vector<PrecisionType>(size);
+
+        out_host_ref = std::vector<OutputType>(size);
```

Comment:
`OutputType` is always complex for fwd transforms so this is over-allocated for real-transforms, I believe: is that intentional?

Reply:
I understand your point being that there is redundant data in the output of the real forward DFT, that we are storing unnecessarily. I think the aim here is to KISS. Whilst that does entail additional computational cost and storage, I don't think this a major concern for the tests.

Reply:
No major concern in the current context indeed and I'm not opposed to keeping it as is, but it might be a desirable change when/if extending testing scope to larger sets of tests and/or much larger test sizes. Maybe we can add a note/comment?

Reply:
Comment added in [a67cc38](https://github.com/oneapi-src/oneMKL/pull/261/commits/a67cc3873f4917e54d349bacf4a25f052917c28c).